aten/src/ATen/native/cuda/layer_norm_kernel.cu

```diff
@@ -0,0 +1,442 @@
+#include <ATen/native/layer_norm.h>
+
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+#include <ATen/Dispatch.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/cuda/detail/IndexUtils.cuh>
+#include <THC/THCDeviceUtils.cuh>
+
+#include <c10/cuda/CUDAMathCompat.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+constexpr int kCUDANumThreads = 256;
+constexpr int kCUDABlockReduceNumThreads = 512;
+constexpr int kColwiseReduceTileSize = 32;
+
+template <typename T>
+__inline__ __device__ T WarpReduceSum(T val) {
+#pragma unroll
+  for (int offset = (C10_WARP_SIZE >> 1); offset > 0; offset >>= 1) {
+    val += WARP_SHFL_DOWN(val, offset);
+  }
+  return val;
+}
+
+template <typename T>
+__inline__ __device__ T BlockReduceSum(T val) {
+  __shared__ T shared[C10_WARP_SIZE];
+  const int lid = threadIdx.x % C10_WARP_SIZE;
+  const int wid = threadIdx.x / C10_WARP_SIZE;
+  val = WarpReduceSum(val);
+  if (lid == 0) {
```

Comment:
you need another __syncthreads here for 2 BlockReduceSum's called one after another, otherwise it's a race condition. Can you please run the tests with cuda-memcheck --racecheck to make sure that there are no other races?

Reply:
Thanks for pointing this out. I changed it to passing shared memory address into BlockReduce then it seems there is no need to do __syncthreads(). I run the tests as well as the benchmark with cuda-memcheck --tool racecheck and there is no racing now.